tempesta_fw/t/unit/test_frang.c

```diff
@@ -0,0 +1,394 @@
+		test_conn_cache = kmem_cache_create("tfw_test_conn_cache",
+				        sizeof(TfwConnection), 0, 0, NULL);
+	BUG_ON(test_conn_cache == NULL);
+	conn = kmem_cache_alloc(test_conn_cache, GFP_ATOMIC);
+	BUG_ON(!conn);
+	tfw_connection_init(conn);
+	return conn;
+}
+
+static int
+req_handler(TfwHttpReq  *req)
+{
+	TfwConnection *conn;
+
+	conn = test_conn_alloc();
+	conn->msg = &req->msg;
+	conn->sk = &mocksock;
+	isk = (struct inet_sock *) (&mocksock);
+	isk->inet_saddr = htonl(in_aton(inet_addr));
+	if (!conn->sk->sk_security) {
+		frang_conn_new(conn->sk);
+	}
+
+	return frang_http_req_handler((void *) conn,
+					       req->msg.skb_list.first, HANDLER_OFF);
+}
+
+static TfwHttpReq *
+get_test_req(const char *req)
+{
+	TfwHttpReq *test_req;
+	static char req_str_copy[PAGE_SIZE]; 
+	int len = strlen(req);
+	BUG_ON(len+1 > sizeof(req_str_copy));
+	strcpy(req_str_copy, req);
+	test_req = test_req_alloc(len);
+	tfw_http_parse_req(test_req, req_str_copy, len);
+	return test_req;
+}
+
+TEST(frang, uri)
+{
+	int res;
+	TfwHttpReq *mockreq;
+
+	mockreq = get_test_req("GET /home/index.html HTTP /1.1\r\n");
+	frang_cfg->http_uri_len = 5;
+	mockreq->frang_st = 3;
+
+	res = req_handler(mockreq);
+	EXPECT_EQ(TFW_BLOCK, res);
+	test_req_free(mockreq);
+}
+
+TEST(frang, req_count)
+{
+	int res;
+	int i;
+	TfwConnection mockconn;
+	unsigned long ts;
+	TfwHttpReq *mockreq;
+	FrangAcc *ra;
+
+	mockreq = get_test_req("GET / HTTP/1.1\r\n");
+	frang_cfg->conn_max = 0;
+	frang_cfg->conn_burst = 0;
+	frang_cfg->conn_rate = 0;
+	frang_cfg->req_rate = 5;
+	isk = (struct inet_sock *) (&mocksock);
+	isk->inet_saddr = htonl(in_aton(inet_addr));
+	mockconn.sk = &mocksock;
+
+	res = frang_conn_new(&mocksock);
+	ra = mockconn.sk->sk_security;
+	ts = jiffies * FRANG_FREQ / HZ;
+	i = ts % FRANG_FREQ;
+	ra->history[i].req = 5;
+	mocksock.sk_security = ra;
+	mockreq->frang_st = 0;
+	res = req_handler (mockreq);
+	EXPECT_EQ(TFW_BLOCK, res);
+
+	frang_cfg->req_rate = 5;
+	frang_cfg->req_burst = 5;
+	ra->history[i].req = 5;
+	mocksock.sk_security = ra;
+	mockreq->frang_st = 0;
+	res = req_handler(mockreq);
+	EXPECT_EQ(TFW_BLOCK, res);
+	test_req_free(mockreq);
+}
+
+TEST(frang, max_conn)
+{
+	int res;
+	int i;
+	TfwHttpReq *mockreq;
+	unsigned long ts;
+
+	mockreq = get_test_req("GET / HTTP/1.1\r\n");
+	frang_cfg->conn_max = 5;
+	isk = (struct inet_sock *) (&mocksock);
+	isk->inet_saddr = htonl(in_aton(inet_addr));
+	res = frang_conn_new(&mocksock);
+	((FrangAcc*)mocksock.sk_security)->conn_curr = 5;
+
+	res = req_handler(mockreq);
+	/*conn_max*/
+	EXPECT_EQ(TFW_BLOCK, res);
+
+	ts = jiffies * FRANG_FREQ / HZ;
+	i = ts % FRANG_FREQ;
+	frang_cfg->conn_max = 0;
+	frang_cfg->conn_rate = 5;
+	((FrangAcc*)mocksock.sk_security)->history[i].conn_new = 5;
+	res = req_handler(mockreq);
+	/*conn_rate */
+	EXPECT_EQ(TFW_BLOCK, res);
+
+	frang_cfg->conn_max = 0;
+	frang_cfg->conn_rate = 0;
+	frang_cfg->conn_burst = 5;
+((FrangAcc*)mocksock.sk_security)->history[i].conn_new = 5;
```

Comment:
It could be incremented in the handler process. I alredy saw "overkills" in a log.